File: Python/Detect_Cycles.py
class Graph:
    def __init__(self, V):
        self.V = V
        self.adjList = [[] for _ in range(V)]

    def addEdge(self, u, v):
        self.adjList[u].append(v)

    def hasCycleUtil(self, v, visited, stack):
        visited[v] = True
        stack[v] = True

        for u in self.adjList[v]:
            if not visited[u]:
                if self.hasCycleUtil(u, visited, stack):
                    return True
            elif stack[u]:
                return True  # Cycle detected

        stack[v] = False
        return False

    def hasCycle(self):
        visited = [False] * self.V
        stack = [False] * self.V

        for v in range(self.V):
            if not visited[v]:
                if self.hasCycleUtil(v, visited, stack):
                    return True

        return False

File: Java/Detect_Cycles.java
import java.util.ArrayList;
import java.util.List;

public class Graph {
    private int V;
    private List<List<Integer>> adjList;

    public Graph(int V) {
        this.V = V;
        adjList = new ArrayList<>(V);
        for (int i = 0; i < V; i++) {
            adjList.add(new ArrayList<>());
        }
    }

    public void addEdge(int u, int v) {
        adjList.get(u).add(v);
    }

    private boolean hasCycleUtil(int v, boolean[] visited, boolean[] stack) {
        visited[v] = true;
        stack[v] = true;

        for (int u : adjList.get(v)) {
            if (!visited[u]) {
                if (hasCycleUtil(u, visited, stack)) {
                    return true;
                }
            } else if (stack[u]) {
                return true; // Cycle detected
            }
        }

        stack[v] = false;
        return false;
    }

    public boolean hasCycle() {
        boolean[] visited = new boolean[V];
        boolean[] stack = new boolean[V];

        for (int v = 0; v < V; v++) {
            if (!visited[v]) {
                if (hasCycleUtil(v, visited, stack)) {
                    return true;
                }
            }
        }

        return false;
    }
}
